File: py/symbolic/minidump.py
"""Minidump processing"""

import io
import shutil
from datetime import datetime

from symbolic._compat import range_type
from symbolic._lowlevel import lib, ffi
from symbolic.utils import RustObject, rustcall, attached_refs, encode_path, \
    encode_str, decode_str, CacheReader

__all__ = ['CallStack', 'FrameInfoMap', 'FrameTrust', 'ProcessState',
           'StackFrame', 'CfiCache']


def _make_frame_trust():
    enums = {}
    by_value = {}

    for attr in dir(lib):
        if attr.startswith('SYMBOLIC_FRAME_TRUST_'):
            name = attr[21:].lower().replace('_', '-')
            value = getattr(lib, attr)
            enums[name] = value
            by_value[value] = name

    enums['by_value'] = by_value
    return type('FrameTrust', (), enums)


FrameTrust = _make_frame_trust()
del _make_frame_trust


class CodeModule(RustObject):
    """Carries information about a code module loaded into the crashed process"""
    __dealloc_func__ = None

    @property
    def id(self):
        """ID of the loaded module containing the instruction"""
        return decode_str(self._objptr.id) or None

    @property
    def addr(self):
        """Address at which the module is loaded in virtual memory"""
        return self._objptr.addr

    @property
    def size(self):
        """Size of the loaded module in virtual memory"""
        return self._objptr.size

    @property
    def name(self):
        """File name of the loaded module's debug file"""
        return decode_str(self._objptr.name)


class StackFrame(RustObject):
    """A single frame in the call stack of a crashed process"""
    __dealloc_func__ = None

    @property
    def return_address(self):
        """The frame's return address as saved in registers or on the stack"""
        return self._objptr.return_address

    @property
    def instruction(self):
        """The frame's program counter location as absolute virtual address"""
        return self._objptr.instruction

    @property
    def trust(self):
        """The confidence with with the instruction pointer was retrieved"""
        return FrameTrust.by_value[self._objptr.trust]

    @property
    def module(self):
        """The code module that defines code for this frame"""
        module = CodeModule._from_objptr(self._objptr.module, shared=True)
        if not module.id and not module.addr and not module.size:
            return None

        attached_refs[module] = self
        return module

    @property
    def registers(self):
        if hasattr(self, '_registers'):
            return self._registers

        self._registers = {}
        for idx in range_type(self._objptr.register_count):
            register = self._objptr.registers[idx]
            name = decode_str(register.name)
            value = decode_str(register.value)
            self._registers[name] = value

        return self._registers


class CallStack(RustObject):
    """A thread of the crashed process"""
    __dealloc_func__ = None

    @property
    def thread_id(self):
        """The id of the thread"""
        return self._objptr.thread_id

    @property
    def frame_count(self):
        """The size of the call stack of this thread"""
        return self._objptr.frame_count

    def frames(self):
        """An iterator over all frames in this call stack"""
        for idx in range_type(self.frame_count):
            yield self.get_frame(idx)

    def get_frame(self, idx):
        """Retrieves the stack frame at the given index (0 is the current frame)"""
        if idx < self.frame_count:
            frame = StackFrame._from_objptr(
                self._objptr.frames[idx], shared=True)
            attached_refs[frame] = self
            return frame
        else:
            raise IndexError("index %d out of bounds %d" %
                             (idx, self.frame_count))


class SystemInfo(RustObject):
    """Information about the CPU and OS on which a minidump was generated."""
    __dealloc_func__ = None

    @property
    def os_name(self):
        """A string identifying the operating system, such as "Windows NT",
        "Mac OS X", or "Linux".  If the information is present in the dump but
        its value is unknown, this field will contain a numeric value.  If
        the information is not present in the dump, this field will be empty.
        """
        return decode_str(self._objptr.os_name)

    @property
    def os_version(self):
        """A string identifying the version of the operating system, such as
        "5.1.2600" or "10.4.8".  The version will be formatted as three-
        component semantic version.  If the dump does not contain this
        information, this field will contain "0.0.0"."""
        return decode_str(self._objptr.os_version)

    @property
    def os_build(self):
        """A string identifying the build of the operating system, such as
        "Service Pack 2" or "8L2127".  If the dump does not contain this
        information, this field will be empty."""
        return decode_str(self._objptr.os_build)

    @property
    def cpu_family(self):
        """A string identifying the basic CPU family, such as "x86" or "ppc".
        If this information is present in the dump but its value is unknown,
        this field will contain a numeric value.  If the information is not
        present in the dump, this field will be empty.  The values stored in
        this field should match those used by MinidumpSystemInfo::GetCPU."""
        return decode_str(self._objptr.cpu_family)

    @property
    def cpu_info(self):
        """A string further identifying the specific CPU, such as
        "GenuineIntel level 6 model 13 stepping 8".  If the information is not
        present in the dump, or additional identifying information is not
        defined for the CPU family, this field will be empty."""
        return decode_str(self._objptr.cpu_info)

    @property
    def cpu_count(self):
        """The number of processors in the system.  Will be greater than one for
        multi-core systems."""
        return self._objptr.cpu_count


class ProcessState(RustObject):
    """State of a crashed process"""
    __dealloc_func__ = lib.symbolic_process_state_free

    @classmethod
    def from_minidump(cls, path, frame_infos=None):
        """Processes a minidump and get the state of the crashed process"""
        frame_infos_ptr = frame_infos._objptr if frame_infos is not None else ffi.NULL
        return ProcessState._from_objptr(
            rustcall(lib.symbolic_process_minidump, encode_path(path), frame_infos_ptr))

    @classmethod
    def from_minidump_buffer(cls, buffer, frame_infos=None):
        """Processes a minidump and get the state of the crashed process"""
        frame_infos_ptr = frame_infos._objptr if frame_infos is not None else ffi.NULL
        return ProcessState._from_objptr(rustcall(
            lib.symbolic_process_minidump_buffer,
            ffi.from_buffer(buffer),
            len(buffer),
            frame_infos_ptr,
        ))

    @property
    def requesting_thread(self):
        """The index of the thread that requested a dump be written in the
        threads vector.  If a dump was produced as a result of a crash, this
        will point to the thread that crashed.  If the dump was produced as
        by user code without crashing, and the dump contains extended Breakpad
        information, this will point to the thread that requested the dump.
        If the dump was not produced as a result of an exception and no
        extended Breakpad information is present, this field will be set to -1,
        indicating that the dump thread is not available."""
        return self._objptr.requesting_thread

    @property
    def timestamp(self):
        """The time-date stamp of the minidump (time_t format)"""
        return self._objptr.timestamp

    @property
    def crashed(self):
        """True if the process crashed, false if the dump was produced outside
        of an exception handler."""
        return self._objptr.crashed

    @property
    def crash_time(self):
        """The UTC time at which the process crashed"""
        if self.timestamp == 0:
            return None
        return datetime.utcfromtimestamp(float(self.timestamp))

    @property
    def crash_address(self):
        """If the process crashed, and if crash_reason implicates memory,
        the memory address that caused the crash.  For data access errors,
        this will be the data address that caused the fault.  For code errors,
        this will be the address of the instruction that caused the fault."""
        return self._objptr.crash_address

    @property
    def crash_reason(self):
        """If the process crashed, the type of crash.  OS- and possibly CPU-
        specific.  For example, "EXCEPTION_ACCESS_VIOLATION" (Windows),
        "EXC_BAD_ACCESS / KERN_INVALID_ADDRESS" (Mac OS X), "SIGSEGV"
        (other Unix)."""
        return decode_str(self._objptr.crash_reason)

    @property
    def assertion(self):
        """If there was an assertion that was hit, a textual representation
        of that assertion, possibly including the file and line at which
        it occurred."""
        return decode_str(self._objptr.assertion)

    @property
    def system_info(self):
        """Returns a weak pointer to OS and CPU information."""
        info = SystemInfo._from_objptr(self._objptr.system_info, shared=True)
        attached_refs[info] = self
        return info

    @property
    def thread_count(self):
        """The number of threads that were running in the crashed process"""
        return self._objptr.thread_count

    def threads(self):
        """An iterator over all threads that were running in the crashed process"""
        for idx in range_type(self.thread_count):
            yield self.get_thread(idx)

    def get_thread(self, idx):
        """Retrieves the thread with the specified index"""
        if idx < self.thread_count:
            stack = CallStack._from_objptr(
                self._objptr.threads[idx], shared=True)
            attached_refs[stack] = self
            return stack
        else:
            raise IndexError("index %d out of bounds %d" %
                             (idx, self.thread_count))

    @property
    def module_count(self):
        return self._objptr.module_count

    def modules(self):
        for idx in range_type(self.module_count):
            yield self.get_module(idx)

    def get_module(self, idx):
        if idx < self.module_count:
            module = CodeModule._from_objptr(
                self._objptr.modules[idx], shared=True)
            attached_refs[module] = self
            return module
        else:
            raise IndexError("index %d out of bounds %d" %
                             (idx, self.module_count))


class FrameInfoMap(RustObject):
    """Stack frame information (CFI) for images"""
    __dealloc_func__ = lib.symbolic_frame_info_map_free

    @classmethod
    def new(cls):
        """Creates a new, empty frame info map"""
        return FrameInfoMap._from_objptr(
            rustcall(lib.symbolic_frame_info_map_new))

    def add(self, id, path):
        """Adds CFI for a code module specified by the `id` argument"""
        self._methodcall(lib.symbolic_frame_info_map_add,
                         encode_str(id), encode_path(path))


# The most recent version for the CFI cache file format
CFICACHE_LATEST_VERSION = rustcall(lib.symbolic_cficache_latest_version)


class CfiCache(RustObject):
    """A cache for call frame information (CFI) to improve native stackwalking"""
    __dealloc_func__ = lib.symbolic_cficache_free

    @classmethod
    def from_path(cls, path):
        """Loads a cficache from a file via mmap."""
        return cls._from_objptr(
            rustcall(lib.symbolic_cficache_from_path, encode_path(path)))

    @classmethod
    def from_object(cls, obj):
        """Creates a cficache from the given object."""
        return cls._from_objptr(
            rustcall(lib.symbolic_cficache_from_object, obj._get_objptr()))

    @property
    def file_format_version(self):
        """Version of the file format."""
        return self._methodcall(lib.symbolic_cficache_get_version)

    @property
    def is_latest_file_format(self):
        """Returns true if this is the latest file format."""
        return self.file_format_version >= CFICACHE_LATEST_VERSION

    def open_stream(self):
        """Returns a stream to read files from the internal buffer."""
        buf = self._methodcall(lib.symbolic_cficache_get_bytes)
        size = self._methodcall(lib.symbolic_cficache_get_size)
        return io.BufferedReader(CacheReader(ffi.buffer(buf, size), self))

        buf = ffi.buffer(self._objptr.bytes, self._objptr.len)
        return io.BufferedReader(CacheReader(buf, self))

    def write_to(self, f):
        """Writes the CFI cache into a file object."""
        shutil.copyfileobj(self.open_stream(), f)

File: cabi/src/minidump.rs
use std::ffi::CStr;
use std::mem;
use std::os::raw::c_char;
use std::slice;
use std::str::FromStr;

use symbolic::common::{byteview::ByteView, types::Arch};
use symbolic::debuginfo::Object;
use symbolic::minidump::cfi::{CfiCache, CFICACHE_LATEST_VERSION};
use symbolic::minidump::processor::{
    CallStack, CodeModule, CodeModuleId, FrameInfoMap, ProcessState, RegVal, StackFrame, SystemInfo,
};

use crate::core::SymbolicStr;
use crate::debuginfo::SymbolicObject;

/// Contains stack frame information (CFI) for images.
pub struct SymbolicFrameInfoMap;

/// Indicates how well the instruction pointer derived during stack walking is trusted.
#[repr(u32)]
pub enum SymbolicFrameTrust {
    None,
    Scan,
    CfiScan,
    Fp,
    Cfi,
    Prewalked,
    Context,
}

/// Carries information about a code module loaded into the process during the crash.
#[repr(C)]
pub struct SymbolicCodeModule {
    pub id: SymbolicStr,
    pub addr: u64,
    pub size: u64,
    pub name: SymbolicStr,
}

/// The CPU register value of a stack frame.
#[repr(C)]
pub struct SymbolicRegVal {
    pub name: SymbolicStr,
    pub value: SymbolicStr,
}

/// Contains the absolute instruction address and image information of a stack frame.
#[repr(C)]
pub struct SymbolicStackFrame {
    pub return_address: u64,
    pub instruction: u64,
    pub trust: SymbolicFrameTrust,
    pub module: SymbolicCodeModule,
    pub registers: *mut SymbolicRegVal,
    pub register_count: usize,
}

impl Drop for SymbolicStackFrame {
    fn drop(&mut self) {
        unsafe {
            Vec::from_raw_parts(self.registers, self.register_count, self.register_count);
        }
    }
}

/// Represents a thread of the process state which holds a list of stack frames.
#[repr(C)]
pub struct SymbolicCallStack {
    pub thread_id: u32,
    pub frames: *mut SymbolicStackFrame,
    pub frame_count: usize,
}

impl Drop for SymbolicCallStack {
    fn drop(&mut self) {
        unsafe {
            Vec::from_raw_parts(self.frames, self.frame_count, self.frame_count);
        }
    }
}

/// OS and CPU information in a minidump.
#[repr(C)]
pub struct SymbolicSystemInfo {
    pub os_name: SymbolicStr,
    pub os_version: SymbolicStr,
    pub os_build: SymbolicStr,
    pub cpu_family: SymbolicStr,
    pub cpu_info: SymbolicStr,
    pub cpu_count: u32,
}

/// State of a crashed process in a minidump.
#[repr(C)]
pub struct SymbolicProcessState {
    pub requesting_thread: i32,
    pub timestamp: u64,
    pub crashed: bool,
    pub crash_address: u64,
    pub crash_reason: SymbolicStr,
    pub assertion: SymbolicStr,
    pub system_info: SymbolicSystemInfo,
    pub threads: *mut SymbolicCallStack,
    pub thread_count: usize,
    pub modules: *mut SymbolicCodeModule,
    pub module_count: usize,
}

impl SymbolicProcessState {
    pub unsafe fn from_process_state(state: &ProcessState) -> Self {
        map_process_state(state)
    }
}

impl Drop for SymbolicProcessState {
    fn drop(&mut self) {
        unsafe {
            Vec::from_raw_parts(self.threads, self.thread_count, self.thread_count);
            Vec::from_raw_parts(self.modules, self.module_count, self.module_count);
        }
    }
}

/// Creates a packed array of mapped FFI elements from a slice.
unsafe fn map_slice<T, S, F>(items: &[T], mut mapper: F) -> (*mut S, usize)
where
    F: FnMut(&T) -> S,
{
    let mut vec = Vec::with_capacity(items.len());
    for item in items {
        vec.push(mapper(item));
    }

    let ptr = vec.as_ptr() as *mut S;
    let len = vec.len();

    mem::forget(vec);
    (ptr, len)
}

/// Creates a packed array of mapped FFI elements from an iterator.
unsafe fn map_iter<T, S, I, F>(items: I, mapper: F) -> (*mut S, usize)
where
    I: Iterator<Item = T>,
    F: Fn(T) -> S,
{
    let mut vec = Vec::with_capacity(items.size_hint().0);
    for item in items {
        vec.push(mapper(item));
    }

    vec.shrink_to_fit();
    let ptr = vec.as_ptr() as *mut S;
    let len = vec.len();

    mem::forget(vec);
    (ptr, len)
}

/// Maps a `CodeModule` to its FFI type.
unsafe fn map_code_module(module: &CodeModule) -> SymbolicCodeModule {
    SymbolicCodeModule {
        id: module
            .id()
            .map(|id| id.to_string().into())
            .unwrap_or_default(),
        addr: module.base_address(),
        size: module.size(),
        name: SymbolicStr::from_string(module.code_file()),
    }
}

/// Maps a pair of register name and value to its FFI type.
unsafe fn map_regval(regval: (&str, RegVal)) -> SymbolicRegVal {
    SymbolicRegVal {
        name: regval.0.into(),
        value: regval.1.to_string().into(),
    }
}

/// Maps a `StackFrame` to its FFI type.
unsafe fn map_stack_frame(frame: &StackFrame, arch: Arch) -> SymbolicStackFrame {
    let empty_module = SymbolicCodeModule {
        id: "".into(),
        addr: 0,
        size: 0,
        name: "".into(),
    };

    let (registers, register_count) =
        map_iter(frame.registers(arch).into_iter(), |r| map_regval(r));

    SymbolicStackFrame {
        instruction: frame.instruction(),
        return_address: frame.return_address(arch),
        trust: mem::transmute(frame.trust()),
        module: frame.module().map_or(empty_module, |m| map_code_module(m)),
        registers,
        register_count,
    }
}

/// Maps a `CallStack` to its FFI type.
unsafe fn map_call_stack(stack: &CallStack, arch: Arch) -> SymbolicCallStack {
    let (frames, frame_count) = map_slice(stack.frames(), |f| map_stack_frame(f, arch));
    SymbolicCallStack {
        thread_id: stack.thread_id(),
        frames,
        frame_count,
    }
}

/// Maps a `SystemInfo` to its FFI type.
unsafe fn map_system_info(info: &SystemInfo) -> SymbolicSystemInfo {
    SymbolicSystemInfo {
        os_name: SymbolicStr::from_string(info.os_name()),
        os_version: SymbolicStr::from_string(info.os_version()),
        os_build: SymbolicStr::from_string(info.os_build()),
        cpu_family: SymbolicStr::from_string(info.cpu_family()),
        cpu_info: SymbolicStr::from_string(info.cpu_info()),
        cpu_count: info.cpu_count(),
    }
}

/// Maps a `ProcessState` to its FFI type.
unsafe fn map_process_state(state: &ProcessState) -> SymbolicProcessState {
    let arch = state.system_info().cpu_arch();
    let (threads, thread_count) = map_slice(state.threads(), |s| map_call_stack(s, arch));
    let (modules, module_count) = map_iter(state.modules().iter(), |m| map_code_module(m));

    SymbolicProcessState {
        requesting_thread: state.requesting_thread(),
        timestamp: state.timestamp(),
        crashed: state.crashed(),
        crash_address: state.crash_address(),
        crash_reason: SymbolicStr::from_string(state.crash_reason()),
        assertion: SymbolicStr::from_string(state.assertion()),
        system_info: map_system_info(state.system_info()),
        threads,
        thread_count,
        modules,
        module_count,
    }
}

ffi_fn! {
    /// Creates a new frame info map.
    unsafe fn symbolic_frame_info_map_new() -> Result<*mut SymbolicFrameInfoMap> {
        let map = Box::into_raw(Box::new(FrameInfoMap::new())) as *mut SymbolicFrameInfoMap;
        Ok(map)
    }
}

ffi_fn! {
    /// Adds the CfiCache for a module specified by the `sid` argument.
    unsafe fn symbolic_frame_info_map_add(
        smap: *const SymbolicFrameInfoMap,
        sid: *const SymbolicStr,
        cficache: *mut SymbolicCfiCache,
    ) -> Result<()> {
        let map = smap as *mut FrameInfoMap<'static>;
        let id = CodeModuleId::from_str((*sid).as_str())?;
        let cache = *Box::from_raw(cficache as *mut CfiCache<'static>);

        (*map).insert(id, cache);
        Ok(())
    }
}

ffi_fn! {
    /// Frees a frame info map object.
    unsafe fn symbolic_frame_info_map_free(smap: *mut SymbolicFrameInfoMap) {
        if !smap.is_null() {
            Box::from_raw(smap as *mut FrameInfoMap<'static>);
        }
    }
}

ffi_fn! {
    /// Processes a minidump with optional CFI information and returns the state
    /// of the process at the time of the crash.
    unsafe fn symbolic_process_minidump(
        path: *const c_char,
        smap: *const SymbolicFrameInfoMap,
    ) -> Result<*mut SymbolicProcessState> {
        let byteview = ByteView::from_path(CStr::from_ptr(path).to_str()?)?;
        let map = if smap.is_null() {
            None
        } else {
            Some(&*(smap as *const FrameInfoMap<'static>))
        };

        let state = ProcessState::from_minidump(&byteview, map)?;
        let sstate = SymbolicProcessState::from_process_state(&state);
        Ok(Box::into_raw(Box::new(sstate)))
    }
}

ffi_fn! {
    /// Processes a minidump with optional CFI information and returns the state
    /// of the process at the time of the crash.
    unsafe fn symbolic_process_minidump_buffer(
        buffer: *const c_char,
        length: usize,
        smap: *const SymbolicFrameInfoMap,
    ) -> Result<*mut SymbolicProcessState> {
        let bytes = slice::from_raw_parts(buffer as *const u8, length);
        let byteview = ByteView::from_slice(bytes);
        let map = if smap.is_null() {
            None
        } else {
            Some(&*(smap as *const FrameInfoMap<'static>))
        };

        let state = ProcessState::from_minidump(&byteview, map)?;
        let sstate = map_process_state(&state);
        Ok(Box::into_raw(Box::new(sstate)))
    }
}

ffi_fn! {
    /// Frees a process state object.
    unsafe fn symbolic_process_state_free(sstate: *mut SymbolicProcessState) {
        if !sstate.is_null() {
            Box::from_raw(sstate);
        }
    }
}

/// Represents a symbolic CFI cache.
pub struct SymbolicCfiCache;

ffi_fn! {
    /// Extracts call frame information (CFI) from an Object.
    unsafe fn symbolic_cficache_from_object(
        sobj: *const SymbolicObject,
    ) -> Result<*mut SymbolicCfiCache> {
        let cache = CfiCache::from_object(&*(sobj as *const Object))?;
        Ok(Box::into_raw(Box::new(cache)) as *mut SymbolicCfiCache)
    }
}

ffi_fn! {
    /// Loads a CFI cache from the given path.
    unsafe fn symbolic_cficache_from_path(path: *const c_char) -> Result<*mut SymbolicCfiCache> {
        let byteview = ByteView::from_path(CStr::from_ptr(path).to_str()?)?;
        let cache = CfiCache::from_bytes(byteview)?;
        Ok(Box::into_raw(Box::new(cache)) as *mut SymbolicCfiCache)
    }
}

ffi_fn! {
    /// Returns the file format version of the CFI cache.
    unsafe fn symbolic_cficache_get_version(scache: *const SymbolicCfiCache) -> Result<u32> {
        let cache = scache as *const CfiCache<'static>;
        Ok((*cache).version())
    }
}

ffi_fn! {
    /// Returns a pointer to the raw buffer of the CFI cache.
    unsafe fn symbolic_cficache_get_bytes(scache: *const SymbolicCfiCache) -> Result<*const u8> {
        let cache = scache as *const CfiCache<'static>;
        Ok((*cache).as_slice().as_ptr())
    }
}

ffi_fn! {
    /// Returns the size of the raw buffer of the CFI cache.
    unsafe fn symbolic_cficache_get_size(scache: *const SymbolicCfiCache) -> Result<usize> {
        let cache = scache as *const CfiCache<'static>;
        Ok((*cache).as_slice().len())
    }
}

ffi_fn! {
    /// Releases memory held by an unmanaged `SymbolicCfiCache` instance.
    unsafe fn symbolic_cficache_free(scache: *mut SymbolicCfiCache) {
        if !scache.is_null() {
            let cache = scache as *mut CfiCache<'static>;
            Box::from_raw(cache);
        }
    }
}

ffi_fn! {
    /// Returns the latest CFI cache version.
    unsafe fn symbolic_cficache_latest_version() -> Result<u32> {
        Ok(CFICACHE_LATEST_VERSION)
    }
}
